As requested by CrimsonShrike, Synthesizer 3.0 is here

DeltaEpsilon7787 opened the pull request:
Have fun maintaining that monstrosity, @CrimsonShrike .

Refer to https://github.com/Baystation12/Baystation12/pull/18298 for details.

DeltaEpsilon7787 committed:
Synthesized Instruments (version 3.0)

DeltaEpsilon7787 committed:
GLOB introduction

Anton-Kr commented:
Jesus christ this is horrifying.

Anton-Kr commented on code/modules/synthesized_instruments/instrument_data/obsolete.dm:
```diff
@@ -0,0 +1,207 @@
+	name="Piano"
+	id="ss13piano"
+	samples=list(\
+"12"='code/modules/synthesized_instruments/samples/obsolete/piano/Cn1.ogg',\
```
This screams of the CS grad meme. There has to be a better way of doing this.

DeltaEpsilon7787 replied:
Doing what? I pretty much copied what's in musician.dm.

Or do you mean definitions being done manually like this? As much as I'd like to use file(), it affects performance greatly.

Anton-Kr replied:
Assigning samples to files without having 90+ lines per instrument. May be a major pain in the ass but perhaps renaming the .ogg files to the appropriate number, and putting it in one for loop. Horrible pseudocode ahead but something like this
```
for(var/i = 12; i< 107; i++)
   tostring(i) = 'code/blahblah/'  + i + '.ogg' 

```

Anton-Kr replied:
It's 100% not functionally necessary but might make readability a bit better. :man_shrugging:

DeltaEpsilon7787 replied:
And this is the issue: you can't do that. 'path/to/file' is calculated compile-time, you can't do anything dynamic here.

DeltaEpsilon7787 replied:
And 90+ per instrument is an exaggeration to be honest. Check other definitions: they have only 5-8 lines per instruments. This is the main selling point of this system in the first place: it's _synthesizes_ sounds, you only need a few samples.

Anton-Kr replied:
Fair enough. Didnt realize that that wouldnt be possible.

ghost commented:
How does this affect server lag, and compile times, compared to the old system?

CrimsonShrike committed:
Finished Glob thingies. Removed debug

DeltaEpsilon7787 commented:
@Rebruiza So vastly different that they are practically incomparable. This system is a superset of the older system.

ghost commented:
:+1:

Commit:
Shitty shit music could easily cause runtimes

Amazingly, this bug has never surfaced before despite how simple it is at its nature.

DeltaEpsilon7787 committed:
This time for real

PsiOmegaDelta commented on code/modules/synthesized_instruments/event_manager.dm:
```diff
@@ -0,0 +1,89 @@
+
+/datum/musical_event/New(datum/sound_player/source, mob/subject, sound/object, time, volume)
+	src.source = source
+	src.subject = subject
```
Not handling the possible deletion of `subject`

DeltaEpsilon7787 replied:
Doesn't matter, tick() is the only procedure actually doing something with subject and it handles null vars.

PsiOmegaDelta replied:
It does matter because it prevents a clean GC of the `subject` instance resulting in a hard-del.

PsiOmegaDelta commented on code/modules/synthesized_instruments/event_manager.dm:
```diff
@@ -0,0 +1,89 @@
+/datum/musical_event/New(datum/sound_player/source, mob/subject, sound/object, time, volume)
+	src.source = source
+	src.subject = subject
+	src.object = object
```
Not nulling any reference vars in `/Destroy()`

DeltaEpsilon7787 replied:
In a slightly convoluted way, all /datum/musical_event are garbage collected by BYOND.
The only thing that works with them is /datum/musical_event_manager and once they are gone from events, they are no longer referenced by anything and become eligible for GC.

PsiOmegaDelta commented on code/modules/synthesized_instruments/song_editor.dm:
```diff
@@ -0,0 +1,104 @@
+
+
+/datum/nano_module/song_editor/New(atom/source, datum/synthesized_song/song)
+	src.host = source
```
Not handling the possible deletion of `source`.
Not nulling the value in `/Destroy()`

DeltaEpsilon7787 replied:
Isn't it handled by nano_module.dm and such?

PsiOmegaDelta replied:
Ah, I see the actual problem here. You're not calling `..()` and the full /New-line should be `/datum/nano_module/song_editor/New(var/host, var/topic_manager, datum/synthesized_song/song)`

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments.dm:
```diff
@@ -0,0 +1,204 @@
+
+
+/obj/structure/synthesized_instrument/Initialize()
+	..()
```
`. = ..()`

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments.dm:
```diff
@@ -0,0 +1,204 @@
+
+/obj/structure/synthesized_instrument/Destroy()
+	qdel(src.player)
+	..()
```
`. = ..()`

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments.dm:
```diff
@@ -0,0 +1,204 @@
+
+
+/obj/item/device/synthesized_instrument/Initialize()
+	..()
```
`. = ..()`

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments.dm:
```diff
@@ -0,0 +1,204 @@
+
+
+/obj/item/device/synthesized_instrument/Destroy()
+	qdel(src.player)
```
`QDEL_NULL(player)`

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments.dm:
```diff
@@ -0,0 +1,204 @@
+
+/obj/item/device/synthesized_instrument/Destroy()
+	qdel(src.player)
+	..()
```
`. = ..()`

PsiOmegaDelta commented on code/modules/synthesized_instruments/sound_player.dm:
```diff
@@ -0,0 +1,108 @@
+	for (var/channel in src.song.free_channels)
+		GLOB.musical_config.free_channels += channel // Deoccupy channels
+	song = null
+	qdel(song)
```
`QDEL_NULL(song)`

PsiOmegaDelta commented on code/modules/synthesized_instruments/sound_player.dm:
```diff
@@ -0,0 +1,108 @@
+	src.stored_locations.Cut()
+	src.actual_instrument = null
+	src.instrument = null
+	sleep(1)
```
Do not `sleep()` in `/Destroy()`

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments.dm:
```diff
@@ -0,0 +1,204 @@
+
+
+/obj/structure/synthesized_instrument/Destroy()
+	qdel(src.player)
```
`QDEL_NULL(player)`

PsiOmegaDelta reviewed (changes requested):
`/obj/structure/synthesized_instrument` and `/obj/item/device/synthesized_instrument` being copy-paste versions of each other is not acceptable.

DeltaEpsilon7787 commented:
@PsiOmegaDelta The only reason why there are two same objects is because /obj/structure/synthesized_instrument is used for static objects on the map (like piano), while /obj/item/device/synthesized_instrument is used for things that actually can be held in hands (like guitar). There is no good and easy way to merge this into one object or without changing some unrelated code. The easiest and most straightforward thing to do is copy and pasting.

Also, check musician.dm and violin.dm and see that they are also essentially same objects with a couple changes.

This copypasta also leads to some ugly code in places where object's type doesn't match variable's type.

For example
`/obj/structure/synthesized_instruments/S = some_instrument`

However, some_instrument can also be a /obj/item/device/synthesized_instrument.
As such, I suggest ignoring this issue as it is unnecessarily complicated to solve and does not affect quality in any way.
There is no class that works both as a /obj/structure and /obj/item/device

EditorRUS commented:
This copypasta also leads to some ugly code in places where object's type doesn't match variable's type.

For example
`/obj/structure/synthesized_instruments/S = some_instrument`

However, some_instrument can also be a /obj/item/device/synthesized_instrument.

DeltaEpsilon7787 committed:
PsiOmegaDelta changes

DeltaEpsilon7787 committed:
PsiOmegaDelta changes

PsiOmegaDelta commented:
All in all much of the code which decides when and how to play sounds to mobs just seems like a much messier version of the `/decl/sound_player` implementation.

PsiOmegaDelta commented:
@EditorRUS Then I suggest you learn how to utilize the power of datums to enable sharing code.
The `/decl/sound_player` type mentioned above is an example of this.

I can state one thing, this will not be merged with such an egregious case of copy-paste.

comma commented on code/modules/synthesized_instruments/real_instruments/Synthesizer/synthesizer.dm:
```diff
@@ -0,0 +1,377 @@
+/obj/structure/synthesized_instrument/synthesizer/proc/compose_code(var/html=0)
+	var/code = ""
+	var/line_number = 1
+	if (src.player:code:len)
```
:s on src vars why

DeltaEpsilon7787 replied:
Ah, this is legacy code. It's disabled for now and should be removed.

DeltaEpsilon7787 commented:
@PsiOmegaDelta This system was written at the time when /decl/sound_player didn't exist and moving it to /decl/sound_player will be _very_ painful. Seems like there wasn't even a concept of a sound manager at the time.
For one thing, /datum/sound_player isn't really the same as /decl/sound_player. No, it's more of a glue object that connects things together. It does not handle actual playback, rather it just applies initial effects such as echo and environment and searches for people to send sounds to. This is also an object you are supposed to override to give your sounds specific effects. /decl/sound_player is not a substitution, but more of a proper way to control channel usage and such.

I'm definitely not going to even attempt hooking it up to /decl/sound_player because it would probably break everything and the sink, but maybe @CrimsonShrike will.

Actual sound control is scheduled in song.dm and executed in event_manager.dm and quite precise control, too. Retarded as it seems, it's done the same way in musician.dm [except for event manager part] which this system is based on. The refactor did remove this absurdity together with /datum/synthesized_song [painstakingly], but sadly, this is an older, but working version without all the improvements I made back then.

Another motive for keeping the system is to avoid dependencies as much as possible. I wish for this thing to be as compatible as possible with any build which is the only reason why I do min(max(A, B), C) instead of Clamp(A, B, C) for example. It's a plug-and-play (pun intended).

chaoko99 commented on code/modules/synthesized_instruments/event_manager.dm:
```diff
@@ -0,0 +1,89 @@
+						event.tick()
+					else left_events += event
+				src.events.Cut()
+				src.events += left_events
```
Make this a timer with a length of 0.

chaoko99 replied:
I meant this, everything south of this spawn.

chaoko99 commented on code/modules/synthesized_instruments/song.dm:
```diff
@@ -0,0 +1,210 @@
+	var/list/allowed_suff = list("b", "n", "#", "s")
+	var/list/note_off_delta = list("a"=91, "b"=91, "c"=98, "d"=98, "e"=98, "f"=98, "g"=98)
+	var/list/lines_copy = src.lines.Copy()
+	spawn()
```
Make this into a timer with a length of 0. These are going to call frequently, so you should use something cheaper.

CrimsonShrike committed:
Removes musCode, fixes

CrimsonShrike committed:
Timers

CrimsonShrike committed:
Datumize instrument UI and interaction, removes duplicate

CrimsonShrike commented:
I am having an issue with using decl/sound_player since this whole system relies on synthesized datums, not directly playing the files as sound_player does. Anyone have an idea on how to proceed?

CrimsonShrike committed:
Temp Rebase

CrimsonShrike committed:
Conversion to sound player. Sounds odd

This may need some tweaks

CrimsonShrike committed:
Enable some env features, tweaks

removes certain unused stuff

chaoko99 commented on code/modules/synthesized_instruments/song.dm:
```diff
@@ -77,25 +80,27 @@
 
+	var/current_volume = Clamp(sound_copy.volume, 0, 100)
+	sound_copy.volume = current_volume //Sanitize volume
+	var/datum/sound_token/token = GLOB.sound_player.PlaySoundDatum(src.player.actual_instrument, src.sound_id, sound_copy, src.player.range, prefer_mute = FALSE)
```
`var/datum/sound_token/token = GLOB.sound_player.PlaySoundDatum(player.actual_instrument, sound_id, sound_copy, player.range, prefer_mute = FALSE)`

chaoko99 reviewed (commented):
I'm just gonna push forward the note that `src.` does nothing, and is just a weird decomp bug.

DeltaEpsilon7787 commented:
@chaoko99 This is actually my personal coding style preference.

This is a way to easily separate local variables from attributes and global variables: I explicitly define the namespace.

global is reserved for when I access global variables (now replaced by GLOB), src is used for attributes and everything else is assumed to be a local variable or a global function.

I suggest adopting this technique on your code base because this verbosity allows to avoid scope issues and such.

Sometimes when function calls turn out too long, I might define a local variable referencing required function.


If you look closely, you will find out that this is everywhere in my original code.

CrimsonShrike commented:
After convertion to new system it seems to sound much worse, ongoing efforts to track it, it's not actually dead

CrimsonShrike commented:
Alright, problems are tracked. Updating sounds doesn't work unless they've been given a channel explicitly, but decl sound player is designed to make this not happen, if anyone has suggestions as to how to proceed  if sound player is mandatory, do tell.

DeltaEpsilon7787 commented:
I will have to criticize decl/sound_player here.

It's **not**, in any way, **supposed** to be used for things like the synthesizer. It's clearly never been intended for this. Making them compatible with each other will require major changes to the player which is definitely out of scope for this pull request.

Here's a list of things that needs to be implemented before there's any hope:

1. Direct sound update. In general this system only updates volume for the sound, but nothing really stops one from implementing sliding pitches, pan changes and such. This is not supported by decl/sound_player well.

2. Explicit channel allocation. Each instrument is given a limited amount of channels to ensure the possibility of multiple instruments working well with each other. This is not implemented which will mean that nothing stops you from taking all the channels for yourself.

3. Custom falloff. Let's be honest here: BYOND's 3D sounds suck some serious balls. They're heavily distorted and inconsistent. This seriously affect audio quality of the play which is not nice. I specifically implemented 2D/3D feature to keep a better option around even if it's slightly less realistic. Also custom falloff to simulate directed sounds and such. None of this is supported by decl/sound_player.


In short, decl/sound_player will require major changes before this system and similar could start to work nicely with each other without any dirty hacks or rushed changes.

MistakeNot4892 commented:
Out of curiosity, how much does this add to the RSC?

Given that it has been open for three months there needs to be a consensus here on whether or not it is acceptable for merge without the /decl/sound changes.

DeltaEpsilon7787 commented:
@PsiOmegaDelta is clearly against any reproduction of functionality and would seem to much rather prefer expanding decl/sound_player instead of having a separate system that repeats some of its functionality.

CrimsonShrike commented:
to clarify, before proceeding, the blocking change is letting decl/sound_player allocate channels for single playing sounds (currently only repeating ones get this) as DeltaEpsilon said. Rest can be considered a feature cut, but this is the one thing system won't work with as it prevents sounds from being updated properly.

PsiOmegaDelta commented:
#22132

PsiOmegaDelta commented on code/modules/synthesized_instruments/globals.dm:
```diff
@@ -0,0 +1,245 @@
+	var/sample
+	var/deviation = 0
+
+	New(sample_file, deviation)
```
Relative pathing.

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments/Synthesizer/synthesizer.dm:
```diff
@@ -0,0 +1,48 @@
+	var/datum/instrument/instruments = list()
+
+
+/obj/structure/synthesized_instrument/synthesizer/New()
```
Use `/Initialize()`

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments/Synthesizer/synthesizer.dm:
```diff
@@ -0,0 +1,48 @@
+		if (!anchored && !isinspace())
+			playsound(src.loc, 'sound/items/Ratchet.ogg', 50, 1)
+			to_chat(usr, "<span class='notice'> You begin to tighten \the [src] to the floor...</span>")
+			if (do_after(user, 20))
```
Need to re-check all pre-conditions after a `do_after()`
In this case the anchor and !inspace() status.

CrimsonShrike replied:
handled

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments/Synthesizer/synthesizer.dm:
```diff
@@ -0,0 +1,48 @@
+		else if(anchored)
+			playsound(src.loc, 'sound/items/Ratchet.ogg', 50, 1)
+			to_chat(usr, "<span class='notice'> You begin to loosen \the [src]'s casters...</span>")
+			if (do_after(user, 40))
```
See above.

PsiOmegaDelta commented on code/modules/synthesized_instruments/sound_player.dm:
```diff
@@ -0,0 +1,64 @@
+	return
+
+/datum/sound_player/proc/shouldStopPlaying(mob/user)
+	return actual_instrument:shouldStopPlaying(user)
```
Do not use colons.

PsiOmegaDelta commented on code/modules/synthesized_instruments/sound_player.dm:
```diff
@@ -0,0 +1,64 @@
+
+
+/datum/sound_player/Destroy()
+	src.song.playing = 0
```
What's with the unnecessary use of `src` literately everywhere?
The only exception is when you have a constructor/`New()` where you're assigning arguments to local vars.

DeltaEpsilon7787 replied:
>This is actually my personal coding style preference.

>This is a way to easily separate local variables from attributes and global variables: I explicitly define the namespace.

>global is reserved for when I access global variables (now replaced by GLOB), src is used for attributes and everything else is assumed to be a local variable or a global function.

PsiOmegaDelta commented on code/modules/synthesized_instruments/song_editor.dm:
```diff
@@ -0,0 +1,105 @@
+			var/num = round(value)
+			if(num > src.song.lines.len || num < 1)
+				return
+			var/content = html_encode(input(usr, "Enter your line: ", "Edit line", src.song.lines[num]) as text|null)
```
Also need to re-check all preconditions after an input.
`if(num > song.lines.len || num < 1)` in this case.

CrimsonShrike replied:
done. Probably

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments.dm:
```diff
@@ -0,0 +1,256 @@
+		if ("import")
+			var/t = ""
+			do
+				t = html_encode(input(usr, "Please paste the entire song, formatted:", text("[]", owner.name), t)  as message)
```
Need to check `CanInteractWith(usr, the_physical_object)` after `input()`/`alert()`
`the_physical_object` is probably `owner` in this case but the Topic handling is the worst mess I've ever seen.

Repeat for every single `input()`/`alert()`

CrimsonShrike replied:
Handled now

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments.dm:
```diff
@@ -0,0 +1,256 @@
+		to_chat(usr, "Non-numeric value was given")
+		return 0
+
+	owner.add_fingerprint(usr)
```
The `..()` call in the calling `/Topic()` has already applied the appropriate fingerprints.

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments.dm:
```diff
@@ -0,0 +1,256 @@
+	if (..())
+		return 1
+
+	return real_instrument.Topic_call(href, href_list, usr)
```
Why do some of these topics call `real_instrument.Topic_call()` unconditionally while others check if `real_instrument` isn't null first?

CrimsonShrike replied:
standarized it, ping me if I missed one

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments/Synthesizer/synthesizer.dm:
```diff
@@ -0,0 +1,48 @@
+		if (!new_instrument.id) continue
+		new_instrument.create_full_sample_deviation_map()
+		src.instruments[new_instrument.name] = new_instrument
+	src.real_instrument = new /datum/real_instrument(src, new /datum/sound_player/synthesizer(src, instruments[pick(instruments)]), instruments)
```
Needs to be `QDEL_NULL()`d in `/Destroy()`

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments/Synthesizer/synthesizer.dm:
```diff
@@ -0,0 +1,48 @@
+/obj/structure/synthesized_instrument/synthesizer/New()
+	..()
+	for (var/type in typesof(/datum/instrument))
+		var/datum/instrument/new_instrument = new type
```
The `/datum/instrument` type appears to be stateless.
Why does every single object that which to use instruments have to create their own instances?

DeltaEpsilon7787 replied:
It's not stateless: create_full_sample_deviation_map mutates it.
Sample map should probably be static though.

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments/Synthesizer/synthesizer.dm:
```diff
@@ -0,0 +1,48 @@
+	..()
+	for (var/type in typesof(/datum/instrument))
+		var/datum/instrument/new_instrument = new type
+		if (!new_instrument.id) continue
```
Use `initial()` to check this var before creating instances unnecessarily.

PsiOmegaDelta commented on code/modules/synthesized_instruments/song.dm:
```diff
@@ -0,0 +1,209 @@
+
+/datum/synthesized_song/proc/play_lines(mob/user, list/allowed_suff, list/note_off_delta, list/lines)
+	Start
+	if (!lines.len) goto Stop
```
`goto`s
I can sense that code smell from all the way over here.

DeltaEpsilon7787 replied:
The original code from which this function originated is even worse. (musician.dm)
`goto` here is used to get out of 3 levels of nested loops instead of using a billion flags.

PsiOmegaDelta commented on code/modules/synthesized_instruments/song.dm:
```diff
@@ -0,0 +1,209 @@
+			cur_note++
+			src.player.event_manager.suspended = 0
+			if (!src.playing || src.player.shouldStopPlaying(user)) goto Stop
+			sleep(duration)
```
Never sleep in a process called proc without setting `set waitfor = 0`

DeltaEpsilon7787 replied:
`waitfor` is deprecated, also this function is called using `addtimer` which already sets up this condition.

PsiOmegaDelta commented on code/modules/synthesized_instruments/song_editor.dm:
```diff
@@ -0,0 +1,105 @@
+
+
+/datum/nano_module/song_editor/New(var/host, var/topic_manager, datum/synthesized_song/song)
+	. = ..()
```
According to the word of God `/New()` shall never return anything.
Don't assign `..()` to `.`

PsiOmegaDelta commented on code/modules/synthesized_instruments/event_manager.dm:
```diff
@@ -0,0 +1,91 @@
+
+
+
+/datum/musical_event/Destroy()
```
Need to call and return `..()`

PsiOmegaDelta commented on code/modules/synthesized_instruments/event_manager.dm:
```diff
@@ -0,0 +1,91 @@
+datum/musical_event
```
Is this type ever created anywhere?

DeltaEpsilon7787 replied:
Yes, push_event

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments.dm:
```diff
@@ -0,0 +1,256 @@
+
+
+/obj/item/device/synthesized_instrument
+	var/datum/real_instrument/real_instrument
```
This var never appears to be initialized.

DeltaEpsilon7787 replied:
It's initialized in subclasses, /obj/item/device/synthesized_instrument is abstract.

PsiOmegaDelta commented on code/modules/synthesized_instruments/usage_info.dm:
```diff
@@ -0,0 +1,40 @@
+/datum/nano_module/usage_info/proc/track_usage()
+	var/cur_channels = src.song_channels.len
+	var/cur_events = src.event_manager_events.len
+	spawn while (src && src.song_channels && src.event_manager_events)
```
Implement `/Process()` and utilize `SSprocessing`
Do not implement your own processing loops.

PsiOmegaDelta commented on code/modules/synthesized_instruments/usage_info.dm:
```diff
@@ -0,0 +1,40 @@
+		var/new_channel_len = round(src.song_channels.len / GLOB.musical_config.usage_info_channel_resolution)
+		var/new_event_len = round(src.event_manager_events.len / GLOB.musical_config.usage_info_event_resolution)
+		if (cur_channels != new_channel_len || cur_events != new_event_len)
+			GLOB.nanomanager.update_uis(src)
```
You can tell NanoUI to auto-refresh itself by setting `ui.set_auto_update(1)` in the `/ui_interact()` setup.
There is no need to do it manually.

PsiOmegaDelta replied:
You've even done so already. The call here is extra unnecessary.

PsiOmegaDelta commented on code/modules/synthesized_instruments/usage_info.dm:
```diff
@@ -0,0 +1,40 @@
+
+/datum/nano_module/usage_info/ui_interact(mob/user, ui_key = "usage_info", var/datum/nanoui/ui = null, var/force_open = 0)
+	var/global/list/data = list()
+	data.Cut()
```
..what? Why are you doing `Cut()` on an already empty list?

DeltaEpsilon7787 replied:
data here is a static variable that is emptied on each interaction.

PsiOmegaDelta commented on code/modules/synthesized_instruments/sound_player.dm:
```diff
@@ -0,0 +1,64 @@
+	src.actual_instrument = null
+	src.instrument = null
+	song = null
+	QDEL_NULL(song)
```
This won't do anything because you nulled `song` above.

PsiOmegaDelta commented on code/modules/synthesized_instruments/real_instruments/Synthesizer/synthesizer.dm:
```diff
@@ -0,0 +1,48 @@
+/obj/structure/synthesized_instrument/synthesizer/New()
+	..()
+	for (var/type in typesof(/datum/instrument))
+		var/datum/instrument/new_instrument = new type
```
But seeing as you currently are creating these instances, when are they destroyed?
I'm not seeing the corresponding `/Destroy()`-proc.

CrimsonShrike commented:
Still working on this, my research project is slowing me down but trying to think of how to use channel stuff, since idea was to keep the whole Decl/sound_player instead of reverting to the system I was told to scrap.

CrimsonShrike committed:
Mass correction

CrimsonShrike commented:
Feel free to slap a big DNM on this until I make sure it works, code correctness aside

CrimsonShrike committed:
new sound datum type, removes some unused code

CrimsonShrike committed:
New instruments, handles some deletion issues

kperdue2002 commented:
would this be compatible with the Aurora codebase? If not, how hard would I have to work to transfer it over, and do I have your permission to do so?

MistakeNot4892 commented:
Aurora and Bay are both AGPL, there is no issue with porting the code.

Compatibility, no idea.
